power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -560,26 +560,14 @@ inline void create_tap_regulator_output(State const& state, std::vector<SolverOu
     }
 }
 
-template <main_core::main_model_state_c State, steady_state_solver_output_type SolverOutputType>
-inline void add_tap_regulator_output(State const& state, TransformerTapRegulator const& regulator, IntS tap_pos,
-                                     std::vector<SolverOutputType>& solver_output) {
-    Idx2D const& math_id =
-        get_math_id<TransformerTapRegulator>(state, get_topology_index<TransformerTapRegulator>(state, regulator.id()));
-    solver_output[math_id.group].transformer_tap_regulator[math_id.pos] = {tap_pos};
-}
-
-template <typename... RegulatedTypes, main_core::main_model_state_c State,
-          steady_state_solver_output_type SolverOutputType>
-    requires(main_core::component_container_c<typename State::ComponentContainer, RegulatedTypes> && ...)
-void add_tap_regulator_output(State const& state,
-                              std::vector<std::vector<TapRegulatorRef<RegulatedTypes...>>> const& regulator_order,
-                              std::vector<SolverOutputType>& solver_output) {
-    create_tap_regulator_output(state, solver_output);
-
-    for (auto const& same_rank_regulators : regulator_order) {
-        for (auto const& regulator : same_rank_regulators) {
-            add_tap_regulator_output(state, regulator.regulator.get(), regulator.transformer.tap_pos(), solver_output);
-        }
+template <transformer_c Component, class ComponentContainer>
+    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component>
+constexpr void get_transformer_tap_positions(main_core::MainModelState<ComponentContainer> const& state,
+                                             TransformerTapPositionResult& transformer_tap_positions) {
+    constexpr auto group_index = ComponentContainer::template get_type_idx<Component>();
+    for (auto const& transformer : state.components.template citer<Component>()) {
+        transformer_tap_positions.emplace_back(
+            TransformerTapPosition{Idx2D{group_index, transformer.id()}, static_cast<IntS>(transformer.tap_pos())});
     }
 }
```

Comment:
as it is now, you produce tap positions for **all** components but the output for the user only concerns the tap positions of regulated transformers.

The way it is now, in the `main_core/output.hpp`, we'll need to loop over all transformers; find the regulator that regulates that transformer (if any); and then output the tap position to the transformer.

That's very expensive and actually not useful because set of regulated transformers is usually fairly small

Reply:
All *transformers*, yes. The argument is that users wouldn't know which transformer is regulated, but they would be interested to know the tap of a transformer (given idx).

I don't see the need to go via the regulators.

Reply:
the user output of the transformer tap regulator is the actual tap position. The user chooses which transformer a regulator regulates, and from there, the regulator is able to tell the user which tap position it chose. This is defined in https://github.com/PowerGridModel/power-grid-model/blob/main/power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/output.hpp#L229

Since outputting this actually requires a regulator-to-transformer or transformer-to-regulator-lookup (expensive), my guess is that it would be easier to go directly via the regulator. Also: the user is not interested in all transformers' tap positions but only the regulated ones, which may be a very small subset of all transformers

Reply:
> the user output of the transformer tap regulator is the actual tap position. 

A regulator is simply a middle man since the tap position is a physical property of transformers. This statement is not correct.

> The user chooses which transformer a regulator regulates, and from there, the regulator is able to tell the user which tap position it chose. 

In the ideal case, a list of transformers to regulate is made known by the users and the same list should be used to retrieve tap positions.

> Since outputting this actually requires a regulator-to-transformer or transformer-to-regulator-lookup (expensive), my guess is that it would be easier to go directly via the regulator. Also: the user is not interested in all transformers' tap positions but only the regulated ones, which may be a very small subset of all transformers

I don't want to argue about efficiency yet, without knowing whether it is actually a bottleneck at all. Saying something is expensive out of the box seems rather contextless. I do agree that there is a subset of what might be interesting for the users, as to which degree I have no knowledge and frankly I don't think there is a guarantee.

As I mentioned above, if a list is made known by the users, we should be using that, instead of going via the regulators.

Reply:
> A regulator is simply a middle man since the tap position is a physical property of transformers. This statement is not correct.

in that case we will need to redesign the C API component output and we need to do it ASAP because if we release the C API with auto tap regulator with tap positions and we want to change that, it will be a breaking change.

Reply:
see e.g. https://power-grid-model--606.org.readthedocs.build/en/606/user_manual/components.html#id38

Reply:
in particular, see this change in the user output data attribute declaration https://github.com/PowerGridModel/power-grid-model/pull/534/files#diff-401bcdd1f75ea6c2e0f22a6b89435d2e0be22fec43a411e5f6f9776a5dd3f8c3R284-R288